consensus/src/txn_manager.rs

```diff
@@ -50,28 +62,40 @@ impl TxnManager for MempoolProxy {
                 });
             }
         }
-        let (callback, callback_rcv) = oneshot::channel();
-        let req = ConsensusRequest::GetBlockRequest(max_size, exclude_txns, callback);
-        // send to shared mempool
-        self.consensus_to_mempool_sender
-            .clone()
-            .try_send(req)
-            .map_err(anyhow::Error::from)?;
-        // wait for response
-        match monitor!(
-            "pull_txn",
-            timeout(Duration::from_secs(1), callback_rcv).await
-        ) {
-            Err(_) => {
-                Err(format_err!("[consensus] did not receive GetBlockResponse on time").into())
+        let no_pending_txns = exclude_txns.is_empty();
+        // keep polling mempool until there's txn available or there's still pending txns
+        for i in 0..self.poll_count {
+            let (callback, callback_rcv) = oneshot::channel();
+            let req = ConsensusRequest::GetBlockRequest(max_size, exclude_txns.clone(), callback);
+            // send to shared mempool
+            self.consensus_to_mempool_sender
+                .clone()
+                .try_send(req)
+                .map_err(anyhow::Error::from)?;
+            // wait for response
+            let txns = match monitor!(
+                "pull_txn",
+                timeout(Duration::from_secs(1), callback_rcv).await
+            ) {
+                Err(_) => Err(MempoolError::from(anyhow::anyhow!(
+                    "[consensus] did not receive GetBlockResponse on time"
+                ))),
+                Ok(resp) => match resp.map_err(anyhow::Error::from)?? {
+                    ConsensusResponse::GetBlockResponse(txns) => Ok(txns),
+                    _ => Err(MempoolError::from(anyhow::anyhow!(
+                        "[consensus] did not receive expected GetBlockResponse"
+                    ))),
+                },
+            }?;
+            if txns.is_empty() && no_pending_txns {
```

Comment:
problem is you're gonna hit this on i == self.poll_count - 1, and you don't want to